### PR TITLE
ec2_vpc_route_table_facts.py migrated to boto3

### LINKED ISSUE #63112: ec2_vpc_route_table_info depends on boto , needs to be upgraded to boto3
##### SUMMARY
Boto is pretty much dead and the upgrade path is boto3

##### COMPONENT NAME

lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_info.py

### PR DESCRIPTION
**Pull request by mpostument**
##### SUMMARY
Migrated to boto3

Fixes #63112

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_vpc_route_table_facts.py

##### ANSIBLE VERSION

```
2.6.3
```

##### ADDITIONAL INFORMATION
```
---
- hosts: localhost
  connection: local
  tasks:
    - ec2_vpc_route_table_facts:
      register: routes

    - debug:
        msg: "{{ routes }}"
```
Boto
```
task path: /home/vagrant/ansible-dev/play.yml:5
ok: [localhost] => {"changed": false, "route_tables": [{"associations": [{"id": "rtbassoc-4c3f7129", "main": true, "route_table_id": "rtb-05641160", "subnet_id": null}], "id": "rtb-05641160", "routes": [{"destination_cidr_block": "172.31.0.0/16", "gateway_id": "local", "instance_id": null, "interface_id": null, "origin": "CreateRouteTable", "state": "active", "vpc_peering_connection_id": null}, {"destination_cidr_block": "0.0.0.0/0", "gateway_id": "igw-1b1cab7e", "instance_id": null, "interface_id": null, "origin": "CreateRoute", "state": "active", "vpc_peering_connection_id": null}], "tags": {}, "vpc_id": "vpc-e5543380"}]}

TASK [debug] *********************************************************************************************************************************************************************************************************************************

task path: /home/vagrant/ansible-dev/play.yml:8
ok: [localhost] => {
    "msg": {
        "changed": false,
        "failed": false,
        "route_tables": [
            {
                "associations": [
                    {
                        "id": "rtbassoc-4c3f7129",
                        "main": true,
                        "route_table_id": "rtb-05641160",
                        "subnet_id": null
                    }
                ],
                "id": "rtb-05641160",
                "routes": [
                    {
                        "destination_cidr_block": "172.31.0.0/16",
                        "gateway_id": "local",
                        "instance_id": null,
                        "interface_id": null,
                        "origin": "CreateRouteTable",
                        "state": "active",
                        "vpc_peering_connection_id": null
                    },
                    {
                        "destination_cidr_block": "0.0.0.0/0",
                        "gateway_id": "igw-1b1cab7e",
                        "instance_id": null,
                        "interface_id": null,
                        "origin": "CreateRoute",
                        "state": "active",
                        "vpc_peering_connection_id": null
                    }
                ],
                "tags": {},
                "vpc_id": "vpc-e5543380"
            }
        ]
    }
}
```
Boto3
```
TASK [ec2_vpc_route_table_facts] *************************************************************************************************************************************************************************************************************

task path: /home/vagrant/ansible-dev/play.yml:5
ok: [localhost] => {"changed": false, "route_tables": [{"associations": [{"main": true, "route_table_association_id": "rtbassoc-4c3f7129", "route_table_id": "rtb-05641160"}], "propagating_vgws": [], "route_table_id": "rtb-05641160", "routes": [{"destination_cidr_block": "172.31.0.0/16", "gateway_id": "local", "origin": "CreateRouteTable", "state": "active"}, {"destination_cidr_block": "0.0.0.0/0", "gateway_id": "igw-1b1cab7e", "origin": "CreateRoute", "state": "active"}], "tags": {}, "vpc_id": "vpc-e5543380"}]}

TASK [debug] *********************************************************************************************************************************************************************************************************************************

task path: /home/vagrant/ansible-dev/play.yml:8
ok: [localhost] => {
    "msg": {
        "changed": false,
        "failed": false,
        "route_tables": [
            {
                "associations": [
                    {
                        "main": true,
                        "route_table_association_id": "rtbassoc-4c3f7129",
                        "route_table_id": "rtb-05641160"
                    }
                ],
                "propagating_vgws": [],
                "route_table_id": "rtb-05641160",
                "routes": [
                    {
                        "destination_cidr_block": "172.31.0.0/16",
                        "gateway_id": "local",
                        "origin": "CreateRouteTable",
                        "state": "active"
                    },
                    {
                        "destination_cidr_block": "0.0.0.0/0",
                        "gateway_id": "igw-1b1cab7e",
                        "origin": "CreateRoute",
                        "state": "active"
                    }
                ],
                "tags": {},
                "vpc_id": "vpc-e5543380"
            }
        ]
    }
}
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py [hunk 1]
```diff
@@ -51,79 +56,175 @@
...
-    for route_table in all_route_tables:
-        route_table_dict_array.append(get_route_table_info(route_table))
-
-    module.exit_json(route_tables=route_table_dict_array)
+        result = json.loads(json.dumps(route_table.describe_route_tables(**params), default=date_handler))
+    except ValueError as e:
+        module.fail_json_aws(e, msg="Cannot validate JSON data")
+    except (ClientError, BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Could not describe route tables")
+    snaked_route_tables = [camel_dict_to_snake_dict(route_table) for route_table in result['RouteTables']]
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py]: I think you might need `camel_dict_to_snake_dict(route_table, ignore_list=['Tags'])`. Looks like CamelCase tags might be modified right now.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py [hunk 2]
```diff
@@ -51,79 +56,175 @@
...
 
-    module = AnsibleModule(argument_spec=argument_spec,
-                           supports_check_mode=True)
-
-    if not HAS_BOTO:
-        module.fail_json(msg='boto required for this module')
-
-    region, ec2_url, aws_connect_params = get_aws_connection_info(module)
+    module = AnsibleAWSModule(argument_spec=argument_spec,
+                              mutually_exclusive=[['vpn_connection_ids', 'filters']],
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py]: s/vpn_connection_ids/route_table_ids/

But do they need to be mutually exclusive? Looks like describe_route_tables takes both at the same time: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_route_tables

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py [hunk 3]
```diff
@@ -51,79 +56,175 @@
...
+                              supports_check_mode=True)
 
-    if region:
-        try:
-            connection = connect_to_aws(boto.vpc, region, **aws_connect_params)
-        except (boto.exception.NoAuthHandlerFound, AnsibleAWSError) as e:
-            module.fail_json(msg=str(e))
-    else:
-        module.fail_json(msg="region must be specified")
+    if not HAS_BOTO3:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py]: AnsibleAWSModule checks this for you so you can remove it.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py [hunk 4]
```diff
@@ -51,79 +56,175 @@
...
-    HAS_BOTO = False
-
-from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import AnsibleAWSError, connect_to_aws, ec2_argument_spec, get_aws_connection_info
-
+    pass  # caught by imported HAS_BOTO3
 
-def get_route_table_info(route_table):
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (ec2_argument_spec, boto3_conn, ansible_dict_to_boto3_filter_list, HAS_BOTO3,
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py]: The boto3_conn and HAS_BOTO3 imports can be removed

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py [hunk 5]
```diff
@@ -24,6 +24,11 @@
     description:
       - A dict of filters to apply. Each dict item consists of a filter key and a filter value.
         See U(http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeRouteTables.html) for possible filters.
+  route_table_ids:
+    description:
+      - Get details of a specific route tables. This value should be provided as a list.
+    required: false
+    version_added: "2.7"
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table_facts.py]: "2.8"

**Review by s-hertel — COMMENTED** [review]
Looks good. This should also have an integration test suite to help maintain backward compatibility (such as for the return values).

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
Would like to see a diff of before/after return values to make sure we won't be breaking any current users for this module.

**Comment by prasadkatti** [comment]
Hi @s-hertel . I would like to work on the integration test suite for this module. Is there a page that describes how to get started with that? Or could you point me to the test suite for a different module that I can look at for reference? I am guessing I will need to setup the test environment in my personal AWS account. Right?

**Pull request closed without merging**